src/core/fieldformatter/qgsvaluerelationfieldformatter.cpp

```diff
@@ -173,43 +174,50 @@ QStringList QgsValueRelationFieldFormatter::valueToStringList( const QVariant &v
   {
     checkList = value.toStringList();
   }
-  else if ( value.type() == QVariant::String )
+  else
   {
-    // This must be an array representation
-    auto newVal { value };
-    if ( newVal.toString().trimmed().startsWith( '{' ) )
-    {
-      newVal = QVariant( newVal.toString().trimmed().mid( 1 ).mid( 0, newVal.toString().length() - 2 ).prepend( '[' ).append( ']' ) );
-    }
-    if ( newVal.toString().trimmed().startsWith( '[' ) )
+    QVariantList valuesList;
+    if ( value.type() == QVariant::String )
     {
-      try
+      // This must be an array representation
+      auto newVal { value };
+      if ( newVal.toString().trimmed().startsWith( '{' ) )
+      {
+        //normal case
+        valuesList = QgsPostgresStringUtils::parseArray( newVal.toString() );
+      }
+      else if ( newVal.toString().trimmed().startsWith( '[' ) )
       {
-        for ( auto &element : json::parse( newVal.toString().toStdString() ) )
+        //fallback, in case it's a json array
+        try
         {
-          if ( element.is_number_integer() )
-          {
-            checkList << QString::number( element.get<int>() );
-          }
-          else if ( element.is_number_unsigned() )
-          {
-            checkList << QString::number( element.get<unsigned>() );
-          }
-          else if ( element.is_string() )
+          for ( auto &element : json::parse( newVal.toString().toStdString() ) )
           {
-            checkList << QString::fromStdString( element.get<std::string>() );
+            if ( element.is_number_integer() )
+            {
+              valuesList.push_back( element.get<int>() );
+            }
+            else if ( element.is_number_unsigned() )
+            {
+              valuesList.push_back( element.get<unsigned>() );
+            }
+            else if ( element.is_string() )
+            {
+              valuesList.push_back( QString::fromStdString( element.get<std::string>() ) );
+            }
           }
         }
-      }
-      catch ( json::parse_error &ex )
-      {
-        qDebug() << QString::fromStdString( ex.what() );
+        catch ( json::parse_error &ex )
+        {
+          qDebug() << QString::fromStdString( ex.what() );
```

Comment:
@signedav is this something that can happen in real life? In this case wouldn't a message log entry be better?

Reply:
Rarely but still possible. In the case that someone used json in string fields to fill up the valuerelation, what would be technically possible since 3.8 - and if this json strings are not formatted correctly.

I'm okay with log as well...

Reply:
Can you do a PR for this on master and we can backport it on top of this one?

Reply:
#31286